File: 6-2-2.java
import java.util.*;

public class Main {
    static class Pair {
        int first;
        int second;

        Pair(int first, int second) {
            this.first = first;
            this.second = second;
        }
    }

    static class Node {
        int vertex;
        long distance;

        Node(int vertex, long distance) {
            this.vertex = vertex;
            this.distance = distance;
        }
    }

    static final long INF = (long) 1e18;

    static int N, X, Y, Z;
    static List<Pair>[] E;
    static boolean[] selected;

    static void dijkstra(int start, int end, List<Long> dist) {
        dist.set(start, 0L);
        PriorityQueue<Node> pq = new PriorityQueue<>((a, b) -> Long.compare(a.distance, b.distance));
        pq.add(new Node(start, 0L));
        while (!pq.isEmpty()) {
            Node node = pq.poll();
            long cost = node.distance;
            int here = node.vertex;
            if (dist.get(here) < cost) continue;
            for (Pair edge : E[here]) {
                int there = edge.first;
                long nextDist = cost + edge.second;
                if (dist.get(there) > nextDist) {
                    dist.set(there, nextDist);
                    pq.add(new Node(there, nextDist));
                }
            }
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int q, u, v, x, i;
        N = sc.nextInt();
        q = sc.nextInt();
        E = new ArrayList[N + 1];
        for (i = 1; i <= N; i++) E[i] = new ArrayList<>();
        while (q-- > 0) {
            u = sc.nextInt();
            v = sc.nextInt();
            x = sc.nextInt();
            E[u].add(new Pair(v, x));
            E[v].add(new Pair(u, x));
        }
        X = sc.nextInt();
        Z = sc.nextInt();
        q = sc.nextInt();
        List<Integer> P = new ArrayList<>();
        for (i = 0; i < q; i++) P.add(sc.nextInt());
        List<Long> dist_X = new ArrayList<>(Collections.nCopies(N + 1, INF));
        selected = new boolean[N + 1];
        dijkstra(X, Z, dist_X);
        List<Long> dist_Z = new ArrayList<>(Collections.nCopies(N + 1, INF));
        selected = new boolean[N + 1];
        dijkstra(Z, X, dist_Z);
        long ans = INF;
        for (int p : P) {
            if (dist_X.get(p) < INF && dist_Z.get(p) < INF && ans > dist_X.get(p) + dist_Z.get(p))
                ans = dist_X.get(p) + dist_Z.get(p);
        }
        if (ans >= INF) ans = -1;
        System.out.println(ans);
    }
}

File: 6-2-2.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef pair<int, int> pii;
const ll INF = ll(1e18);
int N, X, Y, Z;
vector<pii> E[100004];
bool selected[100004];
void dijkstra(int start, int end, vector<ll>& dist) {
	dist[start] = 0;
	priority_queue<pair<ll, int>> pq;
	pq.push({ 0, start });
	while (!pq.empty()) {
		ll cost = -pq.top().first;
		int here = pq.top().second;
		pq.pop();

		if (dist[here] < cost) continue;

		for (int i = 0; i < (int)E[here].size(); ++i) {
			int there = E[here][i].first;
			ll next_dist = cost + E[here][i].second;

			if (dist[there] > next_dist) {
				dist[there] = next_dist;
				pq.push(make_pair(-next_dist, there));
			}
		}
	}
}

int main(void) {
	int q, u, v, x, i;
	scanf("%d%d\n", &N, &q);
	while (q-- > 0) {
		scanf("%d%d%d\n", &u, &v, &x);
		E[u].push_back({ v,x });
		E[v].push_back({ u,x });
	}
	scanf("%d%d\n", &X, &Z);
	scanf("%d\n", &q);
	vector<int> P(q);
	for (auto& p : P)
		scanf("%d", &p);

	vector<ll> dist_X(N + 1, INF);
	for (i = 1; i <= N; i++) selected[i] = false;
	dijkstra(X, Z, dist_X);

	vector<ll> dist_Z(N + 1, INF);
	for (i = 1; i <= N; i++) selected[i] = false;
	dijkstra(Z, X, dist_Z);

	ll ans = INF;
	for (auto p : P) {
		if (dist_X[p] < INF && dist_Z[p]<INF && ans>dist_X[p] + dist_Z[p])
			ans = dist_X[p] + dist_Z[p];
	}

	if (ans >= INF) ans = -1;
	printf("%lld\n", ans);
}